tests/qos/test_buffer.py

```diff
@@ -1888,3 +1959,189 @@ def test_buffer_model_test(duthosts, rand_one_dut_hostname, conn_graph_facts):
         pytest_assert(buffer_model == 'dynamic', 'Got buffer model {} after executing recovering the buffer model to dynamic')
     finally:
         _recovery_to_dynamic_buffer_model(duthost)
+
+
+def calculate_headroom_data(duthost, port_to_test):
+    """
+    This function is intend to calculate the headroom size based on the unput port attributes
+    Each vendor should have it's own implementation for the algorithm
+    """
+    if ASIC_TYPE == 'mellanox':
+        return mellanox_calculate_headroom_data(duthost, port_to_test)
+    else:
+        return False, None
+
+
+def mellanox_calculate_headroom_data(duthost, port_to_test):
+    """
+    This function is Mellanox platform specific.
+    It intends to calculate the headroom size based on the unput port attributes(speed, cable_length, number of lanes..., etc)
+    This algorithm is the same as the implementation in https://github.com/Azure/sonic-swss/blob/master/cfgmgr/buffer_headroom_mellanox.lua
+    """
+    global ASIC_TABLE_KEYS_LOADED
+    global CELL_SIZE
+    global PIPELINE_LATENCY
+    global MAC_PHY_DELAY
+
+    global LOSSLESS_TRAFFIC_PATTERN_KEYS_LOADED
+    global LOSSLESS_MTU
+    global SMALL_PACKET_PERCENTAGE
+
+    over_subscribe_ratio = 0
+    peer_response_time = 0
+    port_mtu = 0
+    gearbox_delay = 0
+    is_8lane = False
+    shp_enabled = False
+
+    head_room_data = {}
+
+    # Init pause_quanta_per_speed_dict
+    pause_quanta_per_speed_dict = {400000: 905, 200000: 453, 100000: 394, 50000: 147, 40000: 118, 25000: 80, 10000: 67,
+                                   1000: 2, 100: 1}
+
+    # Get port speed from config DB
+    # Command: redis-cli -n 4 hget "PORT|Ethernet0" 'speed'
+    port_speed_raw = duthost.shell('redis-cli -n 4 hget "PORT|{}" "speed"'.format(port_to_test))['stdout']
+    if port_speed_raw:
+        port_speed = int(port_speed_raw)
+    else:
+        logging.error("failed to get speed from config db for port {}".format(port_to_test))
+        return False, None
+
+    # Get pause_quanta with port speed from pause_quanta_per_speed_dict
+    if port_speed in pause_quanta_per_speed_dict.keys():
+        pause_quanta = pause_quanta_per_speed_dict[port_speed]
+    else:
+        pause_quanta = 60
```

Comment:
better to get the 'peer_response_time' from ASIC_TABLE and then calculate the pause_quanta instead of hardcoding

Reply:
fixed